components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/recovery/UtxoRecoveryFlow.kt

```
@@ -0,0 +1,321 @@
+                    "Parameters: $from - $until, $duration. Time taken: " +
+                    "${Duration.ofMillis(System.currentTimeMillis() - startTime)}. Exceeded the duration of $duration. " +
+                    "There may be more transactions to recover."
+            )
+        } else if (exceededLastNotarizationTime) {
+            log.info(
+                "Ledger recovery result: $numberOfNotarizedTransactions/$numberOfNotNotarizedTransactions/" +
+                    "$numberOfInvalidTransactions/$numberOfSkippedTransactions (Notarized/Not-notarized/Invalidated/Skipped). " +
+                    "Parameters: $from - $until, $duration. Time taken: " +
+                    "${Duration.ofMillis(System.currentTimeMillis() - startTime)}. Exceeded the duration of " +
+                    "$MAX_DURATION_WITHOUT_SUSPENDING without notarizing a transaction. There may be more transactions to recover."
+            )
+        } else {
+            log.info(
+                "Ledger recovery result: $numberOfNotarizedTransactions/$numberOfNotNotarizedTransactions/" +
+                    "$numberOfInvalidTransactions/$numberOfSkippedTransactions (Notarized/Not-notarized/Invalidated/Skipped). " +
+                    "Parameters: $from - $until, $duration. Time taken: " +
+                    "${Duration.ofMillis(System.currentTimeMillis() - startTime)}."
+            )
+        }
+        return numberOfNotarizedTransactions
+    }
+
+    @Suspendable
+    fun findTransactionsToRecover(): List<SecureHash> {
+        return persistenceService.findTransactionsWithStatusCreatedBeforeTime(
+            TransactionStatus.UNVERIFIED,
+            from,
+            until,
+            QUERY_LIMIT
+        )
+    }
+
+    @Suspendable
+    private fun potentiallyRecoverTransaction(id: SecureHash): RecoveredTransactionResult {
+        val transaction = persistenceService.findSignedTransaction(id, TransactionStatus.UNVERIFIED)
```

Possible we want to keep checking the time at each of these calls - but if these take a long time, stuff is probably broken and probably end up retrying the batch in the flow engine as can't make good progress, but if we can throw an exception if deadline(s) reached, maybe we should?